[17.01] Add `unversioned` keyword to limit mappings to requirements without version

mvdbeek opened the pull request:
This is a followup to #3444 that slightly changes the rules with which Requirements inside a `<requirement/>` tag can be mapped. Since this is a new feature I think it would be good to target the post-freeze 17.01 branch.

The new rule is that we do not match a specific requirement (i.e package 0.1.19) when
mapping using `unversioned: true` (Thanks @jmchilton for that).
```
- from:
    name: package
    unversioned: true
  to:
    name: package
    version: 1.3.1
```

Finally I also added an additional set operation on the tool requirements after expanding them using the mapping file (14529b9). That way we avoid creating the `mulled-v1-<hash>` environment if we simply map dependency names. Mapping `sickle` to `sickle-trim` benefits from this.

mvdbeek committed:
Allow mapping requirements with version None to a specific version (so that all-in-one dependency resolution is applied)

but do not match a specific requirement (i.e samtools 0.1.19) when
mapping
```
- from:
    name: samtools
    version: None
  to:
    name: samtools
    version: 1.3.1
```
This will only map samtools requirements that do not specify a version.
If you want to override a specifically requested version, you need to
include that version, i.e
```
- from:
    name: samtools
    version: None
  to:
    name: samtools
    version: 1.3.1
- from:
    name: samtools
    version: 0.1.19
  to:
    name: samtools
    version: 1.2
```

mvdbeek committed:
Extend default mapping

mvdbeek committed:
Use Null instead of None

to get the non-string None value.
str(from_raw.get(version)) will turn it into 'None' anyway.

mvdbeek committed:
Add unit tests that asserts new dependency mapping rules

bgruening commented on config/local_conda_mapping.yml.sample:
```diff
@@ -1,2 +1,50 @@
+    version: 4.6
+- from:
+    name: openbabel
+    version: 2.3.2
```
Why is this mapping in place?

mvdbeek replied:
This specifically because there is no openbabel 2.3.2 on conda.
In which case we might want to upgrade the datatype converters that use that version ?!
Or create a openbabel 2.3.2 package?
Fyi those are the ~~tools~~ converters that want openbabel 2.3.2:
```
cml_to_inchi_converter.xml
cml_to_mol2_converter.xml
cml_to_sdf_converter.xml
cml_to_smi_converter.xml
inchi_to_cml_converter.xml
inchi_to_mol2_converter.xml
inchi_to_mol_converter.xml
inchi_to_sdf_converter.xml
inchi_to_smi_converter.xml
mol2_to_cml_converter.xml
mol2_to_inchi_converter.xml
mol2_to_mol_converter.xml
mol2_to_sdf_converter.xml
mol2_to_smi_converter.xml
mol_to_cml_converter.xml
mol_to_inchi_converter.xml
mol_to_mol2_converter.xml
mol_to_smi_converter.xml
sdf_to_cml_converter.xml
sdf_to_inchi_converter.xml
sdf_to_mol2_converter.xml
sdf_to_smi_converter.xml
smi_to_cml_converter.xml
smi_to_inchi_converter.xml
smi_to_mol2_converter.xml
smi_to_mol_converter.xml
smi_to_sdf_converter.xml
smi_to_smi_converter.xml
```

bgruening replied:
There should be a 2.3.2 version: https://anaconda.org/bioconda/openbabel/files
- but only for Linux. In theory the 2.4 should work, but I think I need to test this before I feel really good about it.

mvdbeek replied:
I can check of I get 2.3.2 to compile on osx ... we do have a cairo library nowadays.

mvdbeek committed:
Drop openbabel mapping from 2.32 to 2.4.1 (thx @bgruening)

mvdbeek committed:
Collapse list of tool requirements after mapping

With this change, the merged environment name for requirements that have
been specified using multiple names (e.g. sickle (for TS) and
sickle-trim (conda/docker)) will be sickle-trim if mapping sickle to
sickle-trim (instead of mulled-v1-<hash>).

jmchilton commented on lib/galaxy/tools/deps/resolvers/conda.py:
```diff
@@ -146,15 +146,11 @@ def resolve_all(self, requirements, **kwds):
-                version = None
-
-            conda_targets.append(CondaTarget(requirement.name, version=version))
+        expanded_requirements = set([self._expand_requirement(r) for r in requirements])
```
I'd prefer a version of this that preserved the order - I think the order might be important for the hashing right?

mvdbeek replied:
oh, yes, that's right. I'll just use the ToolRequirements class, which stores the requirements as an ordered set.

jmchilton replied:
Awesome thanks! I'm about to open a PR against your branch that shuffles some other stuff in the PR around quite a bit but it won't touch this line I don't believe.

mvdbeek committed:
Use ToolRequirements (uses OrderedSet internally)

to collapse redundant requirements instead of set to maintain the order
of specified requirements.

jmchilton committed:
Modify representation of mapping to None. (#4)

I feel that mapping from

```
- from: blast+
  to: blast
```

should be the same as

```
- from:
   name: blast+
   version: Null
  to: blast
```

since it should just be syntatic sugar. And I think that second version reads like it should by default apply to all versions.

I feel therefore it is probably better to explicitly require stating that a mapping only applies to unversioned requirements. This commit does that by introducing the syntax:

```
- from:
    name: blast+
    unversioned: true
  to:
    name: blast
    version: 1.3.4
```

It feels better to have strongly typed entities like this than to introduce magic constants such as:

```
- from:
   name: blast+
   version: __unversioned__
  to:
   name: blast
   version: 1.3.4
```

which is another route this could have gone I guess.

jmchilton commented:
Cool beans - I like everything about this. Thanks @mvdbeek and @bgruening.

nsoranzo commented on lib/galaxy/tools/deps/resolvers/default_conda_mapping.yml:
```diff
@@ -2,4 +2,46 @@
+    unversioned: true
+  to:
+    name: ucsc_tools
+    version: 332
```
@mvdbeek I'm not able to find this package in our default channels, only in your Conda channel.

mvdbeek replied:
I guess this is stuck in https://github.com/galaxyproject/conda-iuc/pull/9/files, I have closed the bioconda PR that was going to add this.
